apps/studio/src/utils/sitemap.ts

```diff
@@ -0,0 +1,101 @@
+import type { IsomerSitemap } from "@opengovsg/isomer-components"
+import type { Resource } from "@prisma/client"
+
+type ResourceDto = Omit<
+  Resource,
+  "id" | "parentId" | "publishedVersionId" | "draftBlobId"
+> & {
+  id: string
+  parentId: string | null
+}
+
+const getSitemapTreeFromArray = (
+  resources: ResourceDto[],
+  parentId: string | null,
+  path: string,
+): IsomerSitemap[] | undefined => {
+  if (resources.length === 0) {
+    return undefined
+  }
+
+  const children = resources.filter((resource) => {
+    if (parentId === null) {
+      return resource.parentId === null && resource.permalink !== ""
+    }
+    return resource.parentId === parentId
+  })
+
+  // Filter out children that have both the Page and the Folder types and keep
+  // only the one with the Folder type, then reassign the ID to the ID of the
+  // resource with the Page type
+  return children
+    .filter(
+      (child) =>
+        !(
+          child.type === "Page" &&
+          children.some(
+            (otherChild) =>
+              otherChild.id !== child.id &&
+              otherChild.permalink === child.permalink &&
+              otherChild.type === "Folder",
+          )
+        ),
+    )
```

Comment:
can help me understand what this filter is trying to accomplish? at first glance, this is trying to filter out pages that have the same permalink as another folder? but not entirely sure why this filter is required here

Reply:
Basically right now we always have duplicate resources for the same permalink when it is a folder. However, we don't have such a representation in the sitemap because the SOT is the permalink. So this filter is trying to remove Pages and CollectionPages that have a corresponding Folder or Collection, then re-assigning the resource ID of the Folder or Collection to the ID that was removed.

The reason we do this is because in the actual sitemap, we only are interested in IDs of actual pages (folders is not a concept that exist in the final site). However, our parent-child relationships uses folders (and collections), so we need to perform this re-assignment to ensure that everything in the generated sitemap are actual pages, and the parent-child relationships are still preserved.